src/lib/operators/join_sort_merge.cpp

```diff
@@ -910,31 +891,74 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractJoinOperatorImpl {
+        null_output_right->insert(null_output_right->end(), _null_rows_left->size(), NULL_ROW_ID);
+        for (const auto& row_id_left : *_null_rows_left) {
+          null_output_left->push_back(row_id_left);
+        }
+      }
+      if (include_null_right) {
+        null_output_left->insert(null_output_left->end(), _null_rows_right->size(), NULL_ROW_ID);
+        null_output_right->reserve(_null_rows_right->size());
+        for (const auto& row_id_right : *_null_rows_right) {
+          null_output_right->push_back(row_id_right);
+        }
+      }
+
+      DebugAssert(null_output_left->size() == null_output_right->size(),
+                  "Null positions lists are expected to be of equal length.");
+      if (!null_output_left->empty()) {
+        _output_pos_lists_left.push_back(null_output_left);
+        _output_pos_lists_right.push_back(null_output_right);
       }
     }
-    if (include_null_right) {
-      for (auto row_id_right : *_null_rows_right) {
-        output_left->push_back(NULL_ROW_ID);
-        output_right->push_back(row_id_right);
+
+    // Intermediate structure for output chunks (to avoid concurrent appending to table)
+    std::vector<std::shared_ptr<Chunk>> output_chunks(_output_pos_lists_left.size());
+
+    // Determine if writing output in parallel is necessary.
+    // As partitions ought to be roughly equally sized, looking at the first should be sufficient.
+    const auto write_output_concurrently = _cluster_count > 1 && _output_pos_lists_left[0]->size() > 10'000;
+
+    std::vector<std::shared_ptr<AbstractTask>> output_jobs;
+    output_jobs.reserve(_output_pos_lists_left.size());
+    for (auto pos_list_id = size_t{0}; pos_list_id < _output_pos_lists_left.size(); ++pos_list_id) {
+      if (_output_pos_lists_left[pos_list_id]->empty() && _output_pos_lists_right[pos_list_id]->empty()) {
+        continue;
+      }
+
+      auto write_output_chunk = [this, pos_list_id, &output_chunks] {
+        Segments segments;
+        _add_output_segments(segments, _sort_merge_join.input_table_left(), _output_pos_lists_left[pos_list_id]);
+        _add_output_segments(segments, _sort_merge_join.input_table_right(), _output_pos_lists_right[pos_list_id]);
+        output_chunks[pos_list_id] = std::make_shared<Chunk>(std::move(segments));
+      };
+
+      if (write_output_concurrently) {
+        auto job = std::make_shared<JobTask>(write_output_chunk);
+        output_jobs.push_back(job);
+        output_jobs.back()->schedule();
+      } else {
+        write_output_chunk();
       }
     }
 
-    // Add the segments from both input tables to the output
-    Segments output_segments;
-    _add_output_segments(output_segments, _sort_merge_join.input_table_left(), output_left);
-    _add_output_segments(output_segments, _sort_merge_join.input_table_right(), output_right);
+    // Wait for all chunk creation tasks to finish
+    Hyrise::get().scheduler()->wait_for_tasks(output_jobs);
```

Comment:
don't know how simple the logic behind wait_for_tasks is but could we avoid calling it by checking write_output_concurrently before?

Reply:
@mrks : do you have an idea about the impact of `wait_for_tasks()`. Given that SMJ is the fastest join for very small data sets, it might make sense to do the check here (even though we will not switch to SMJs in the near future).

Reply:
It just iterates over the input vector. If the vector is empty, it should have a negligible cost. I still like checking for `write_output_concurrently` because it reminds readers that this is not always done.